File: app/lib/data/models/entity_model.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'entity_model.freezed.dart';
part 'entity_model.g.dart';

enum ContentType { article, video, podcast, document, book }

enum ContentStatus { pending, processing, completed, failed }

enum DifficultyLevel { beginner, intermediate, advanced, expert }

@freezed
class ContentMetadata with _$ContentMetadata {
  const factory ContentMetadata({
    String? author,
    DateTime? publishDate,
    int? readingTime,
    @Default(DifficultyLevel.intermediate) DifficultyLevel difficulty,
    required String sourceUrl,
    String? thumbnail,
    int? fileSize,
    int? duration,
    @Default('en') String language,
  }) = _ContentMetadata;

  factory ContentMetadata.fromJson(Map<String, dynamic> json) =>
      _$ContentMetadataFromJson(json);
}

@freezed
class ProcessedContent with _$ProcessedContent {
  const factory ProcessedContent({
    String? summary,
    @Default([]) List<String> keyPoints,
    @Default([]) List<String> tags,
    String? thumbnail,
    @Default([]) List<String> concepts,
    @Default(DifficultyLevel.intermediate) DifficultyLevel difficulty,
    int? readingTime,
  }) = _ProcessedContent;

  factory ProcessedContent.fromJson(Map<String, dynamic> json) =>
      _$ProcessedContentFromJson(json);
}

@freezed
class Flashcard with _$Flashcard {
  const factory Flashcard({
    required String id,
    required String front,
    required String back,
    DifficultyLevel? difficulty,
    String? imageUrl,
    @Default(0) int reviewCount,
    DateTime? lastReviewedAt,
    DateTime? nextReviewAt,
    @Default(2.5) double easeFactor,
  }) = _Flashcard;

  factory Flashcard.fromJson(Map<String, dynamic> json) =>
      _$FlashcardFromJson(json);
}

@freezed
class QuizQuestion with _$QuizQuestion {
  const factory QuizQuestion({
    required String id,
    required String question,
    @Default([]) List<String> options,
    required int correctAnswer,
    String? explanation,
    @Default(DifficultyLevel.intermediate) DifficultyLevel difficulty,
    @Default(10) int points,
  }) = _QuizQuestion;

  factory QuizQuestion.fromJson(Map<String, dynamic> json) =>
      _$QuizQuestionFromJson(json);
}

@freezed
class QuizAttempt with _$QuizAttempt {
  const factory QuizAttempt({
    required int score,
    required int totalQuestions,
    required int correctAnswers,
    int? timeSpent,
    required DateTime completedAt,
    @Default([]) List<int> answers,
  }) = _QuizAttempt;

  factory QuizAttempt.fromJson(Map<String, dynamic> json) =>
      _$QuizAttemptFromJson(json);
}

@freezed
class LearningMaterials with _$LearningMaterials {
  const factory LearningMaterials({
    @Default([]) List<Flashcard> flashcards,
    Quiz? quiz,
    @Default([]) List<String> learningPath,
  }) = _LearningMaterials;

  factory LearningMaterials.fromJson(Map<String, dynamic> json) =>
      _$LearningMaterialsFromJson(json);
}

@freezed
class Quiz with _$Quiz {
  const factory Quiz({
    @Default([]) List<QuizQuestion> questions,
    @Default([]) List<QuizAttempt> attempts,
  }) = _Quiz;

  factory Quiz.fromJson(Map<String, dynamic> json) => _$QuizFromJson(json);
}

@freezed
class UserInteractions with _$UserInteractions {
  const factory UserInteractions({
    @Default(false) bool isRead,
    @Default(false) bool isFavorite,
    int? rating,
    DateTime? readAt,
  }) = _UserInteractions;

  factory UserInteractions.fromJson(Map<String, dynamic> json) =>
      _$UserInteractionsFromJson(json);
}

@freezed
class EntityModel with _$EntityModel {
  const EntityModel._();

  const factory EntityModel({
    required String id,
    required String url,
    required String title,
    required String description,
    required String userId,
    @Default([]) List<String> tags,
    @Default(ContentType.article) ContentType type,
    @Default(ContentStatus.pending) ContentStatus status,
    required ContentMetadata metadata,
    ProcessedContent? processedContent,
    LearningMaterials? learningMaterials,
    @Default(UserInteractions()) UserInteractions userInteractions,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _EntityModel;

  factory EntityModel.fromJson(Map<String, dynamic> json) {
    return EntityModel(
      id: json['_id']?.toString() ?? json['id'] ?? '',
      url: json['url'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      userId: json['userId'] ?? '',
      tags: json['tags'] != null ? List<String>.from(json['tags']) : [],
      type: ContentType.values.firstWhere(
        (e) => e.name == json['type'],
        orElse: () => ContentType.article,
      ),
      status: ContentStatus.values.firstWhere(
        (e) => e.name == json['status'],
        orElse: () => ContentStatus.pending,
      ),
      metadata: json['metadata'] != null
          ? ContentMetadata.fromJson(json['metadata'])
          : ContentMetadata(sourceUrl: json['url'] ?? ''),
      processedContent: json['processedContent'] != null
          ? ProcessedContent.fromJson(json['processedContent'])
          : null,
      learningMaterials: json['learningMaterials'] != null
          ? LearningMaterials.fromJson(json['learningMaterials'])
          : null,
      userInteractions: json['userInteractions'] != null
          ? UserInteractions.fromJson(json['userInteractions'])
          : const UserInteractions(),
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : DateTime.now(),
    );
  }

  factory EntityModel.empty() => EntityModel(
    id: '',
    url: '',
    title: '',
    description: '',
    userId: '',
    metadata: const ContentMetadata(
      sourceUrl: '',
    ),
    createdAt: DateTime.now(),
    updatedAt: DateTime.now(),
  );
}

// Extension methods for computed properties
extension EntityModelX on EntityModel {
  bool get isProcessed => status == ContentStatus.completed;

  bool get isProcessing => status == ContentStatus.processing;

  bool get isFailed => status == ContentStatus.failed;

  List<String> get displayTags => tags.isNotEmpty ? tags : (processedContent?.tags.isNotEmpty == true ? processedContent!.tags : []);

  String get displaySummary =>
      processedContent?.summary ?? description;

  String get domain {
    try {
      final uri = Uri.parse(url);
      return uri.host.replaceAll('www.', '').split('.')[0];
    } catch (_) {
      return 'Unknown';
    }
  }
}

File: services/src/database/models/Entity.model.ts
import mongoose, { Schema, Document, Model } from 'mongoose';
import { Entity, ContentType, ContentStatus, DifficultyLevel } from '../../shared/interfaces/content.interface';

interface EntityDocument extends Omit<Entity, '_id'>, Document {}

const contentMetadataSchema = new Schema({
  author: { type: String },
  publishDate: { type: Date },
  readingTime: { type: Number }, // in minutes
  difficulty: {
    type: String,
    enum: ['beginner', 'intermediate', 'advanced', 'expert'],
    default: 'intermediate'
  },
  sourceUrl: { type: String, required: true },
  thumbnail: { type: String },
  fileSize: { type: Number },
  duration: { type: Number }, // for video/podcast in seconds
  language: { type: String, default: 'en' }
});

const processedContentSchema = new Schema({
  summary: { type: String },
  keyPoints: [{ type: String }],
  tags: [{ type: String }],
  thumbnail: { type: String },
  concepts: [{ type: String }],
  difficulty: {
    type: String,
    enum: ['beginner', 'intermediate', 'advanced', 'expert']
  },
  readingTime: { type: Number }
});

const flashcardSchema = new Schema({
  id: { type: String, required: true },
  front: { type: String, required: true },
  back: { type: String, required: true },
  difficulty: {
    type: String,
    enum: ['beginner', 'intermediate', 'advanced', 'expert']
  },
  imageUrl: { type: String },
  reviewCount: { type: Number, default: 0 },
  lastReviewedAt: { type: Date },
  nextReviewAt: { type: Date },
  easeFactor: { type: Number, default: 2.5 }
});

const quizQuestionSchema = new Schema({
  id: { type: String, required: true },
  question: { type: String, required: true },
  options: [{ type: String, required: true }],
  correctAnswer: { type: Number, required: true },
  explanation: { type: String },
  difficulty: {
    type: String,
    enum: ['beginner', 'intermediate', 'advanced', 'expert']
  },
  points: { type: Number, default: 10 }
});

const quizAttemptSchema = new Schema({
  score: { type: Number, required: true },
  totalQuestions: { type: Number, required: true },
  correctAnswers: { type: Number, required: true },
  timeSpent: { type: Number }, // in seconds
  completedAt: { type: Date, default: Date.now },
  answers: [{ type: Number }]
});

const learningMaterialsSchema = new Schema({
  flashcards: [flashcardSchema],
  quiz: {
    questions: [quizQuestionSchema],
    attempts: [quizAttemptSchema]
  },
  learningPath: [{ type: String }]
});

const userInteractionsSchema = new Schema({
  isRead: { type: Boolean, default: false },
  isFavorite: { type: Boolean, default: false },
  rating: { type: Number, min: 1, max: 5 },
  readAt: { type: Date }
});

const entitySchema = new Schema(
  {
    url: {
      type: String,
      required: true,
      index: true
    },
    title: {
      type: String,
      required: true,
      trim: true
    },
    description: {
      type: String,
      trim: true
    },
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true
    },
    tags: {
      type: [String],
      default: [],
      index: true
    },
    type: {
      type: String,
      enum: ['article', 'video', 'podcast', 'document', 'book'],
      required: true,
      default: 'article'
    },
    status: {
      type: String,
      enum: ['pending', 'processing', 'completed', 'failed'],
      required: true,
      default: 'pending',
      index: true
    },
    metadata: {
      type: contentMetadataSchema,
      required: true
    },
    processedContent: {
      type: processedContentSchema
    },
    learningMaterials: {
      type: learningMaterialsSchema
    },
    userInteractions: {
      type: userInteractionsSchema,
      default: {}
    }
  },
  {
    timestamps: true,
    toJSON: {
      virtuals: true
    },
    toObject: {
      virtuals: true
    }
  }
);

// Indexes
entitySchema.index({ userId: 1, createdAt: -1 });
entitySchema.index({ tags: 1, createdAt: -1 });
entitySchema.index({ status: 1 });
entitySchema.index({ type: 1 });
entitySchema.index({ 'userInteractions.isRead': 1 });
entitySchema.index({ 'userInteractions.isFavorite': 1 });
entitySchema.index({ title: 'text', description: 'text', 'processedContent.summary': 'text' });

// Method to mark as read
entitySchema.methods.markAsRead = function () {
  this.userInteractions.isRead = true;
  this.userInteractions.readAt = new Date();
  return this.save();
};

// Method to toggle favorite
entitySchema.methods.toggleFavorite = function () {
  this.userInteractions.isFavorite = !this.userInteractions.isFavorite;
  return this.save();
};

// Method to add quiz attempt
entitySchema.methods.addQuizAttempt = function (attempt: any) {
  if (!this.learningMaterials) {
    this.learningMaterials = { flashcards: [], quiz: { questions: [], attempts: [] } };
  }
  if (!this.learningMaterials.quiz.attempts) {
    this.learningMaterials.quiz.attempts = [];
  }
  this.learningMaterials.quiz.attempts.push(attempt);
  return this.save();
};

// Static method to find by user
entitySchema.statics.findByUser = function (userId: string) {
  return this.find({ userId }).sort({ createdAt: -1 });
};

// Static method to find pending processing
entitySchema.statics.findPendingProcessing = function () {
  return this.find({ status: 'pending' }).limit(10);
};

// Static method to find by tags
entitySchema.statics.findByTags = function (userId: string, tags: string[]) {
  return this.find({ userId, tags: { $in: tags } }).sort({ createdAt: -1 });
};

interface EntityModel extends Model<EntityDocument> {
  findByUser(userId: string): Promise<EntityDocument[]>;
  findByTags(userId: string, tags: string[]): Promise<EntityDocument[]>;
  findPendingProcessing(): Promise<EntityDocument[]>;
}

const EntityModel = mongoose.model<EntityDocument, EntityModel>('Entity', entitySchema);

export default EntityModel;
